File: json/bench/bench.ts
// Run with node or bun, from this folder, after run.py has written out/doc.json.
import { readFileSync } from "node:fs";

const TEXT = readFileSync("out/doc.json", "latin1");

function chk(s: string): number {
  let h = 0;
  for (let i = 0; i < s.length; i++) h = (Math.imul(h, 31) + s.charCodeAt(i)) >>> 0;
  return (h + s.length) >>> 0;
}

let t0 = performance.now();
const v = JSON.parse(TEXT);
let ms = performance.now() - t0;
console.log(`parse\t${ms.toFixed(3)}\t${chk(JSON.stringify(v))}`);

t0 = performance.now();
const out = JSON.stringify(v);
ms = performance.now() - t0;
console.log(`encode\t${ms.toFixed(3)}\t${chk(out)}`);

File: json/bench/bench.py
import json, time

with open("out/doc.json", encoding="latin-1") as f:
    TEXT = f.read()


def chk(s: str) -> int:
    h = 0
    for c in s:
        h = (h * 31 + ord(c)) & 0xFFFFFFFF
    return (h + len(s)) & 0xFFFFFFFF


def enc(v) -> str:
    return json.dumps(v, separators=(",", ":"), ensure_ascii=False)


t0 = time.perf_counter()
v = json.loads(TEXT)
ms = (time.perf_counter() - t0) * 1000
print(f"parse\t{ms:.3f}\t{chk(enc(v))}")

t0 = time.perf_counter()
out = enc(v)
ms = (time.perf_counter() - t0) * 1000
print(f"encode\t{ms:.3f}\t{chk(out)}")
